custom/icds/utils/ccz_hosting.py

```diff
@@ -0,0 +1,55 @@
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+from memoized import memoized
+from django.urls import reverse
+
+from corehq.blobs import (
+    get_blob_db,
+    CODES,
+)
+
+
+class CCZHostingUtility:
+    def __init__(self, file_hosting):
+        self.domain = file_hosting.domain
+        self.file_hosting = file_hosting
+        self.blob_id = file_hosting.blob_id
+
+    def file_exists(self):
+        return get_blob_db().exists(key=self.blob_id)
+
+    def get_file(self):
+        return get_blob_db().get(key=self.blob_id)
+
+    def get_file_size(self):
+        return get_blob_db().size(key=self.blob_id)
+
+    @memoized
+    def get_file_meta(self):
+        if self.file_exists():
+            return get_blob_db().metadb.get(key=self.blob_id, parent_id='CCZHosting')
+
+    def get_file_name(self):
+        return self.get_file_meta().name if self.get_file_meta() else ''
+
+    @property
+    def ccz_details(self):
+        return {
+            'name': self.file_hosting.file_name,
+            'download_url': reverse('ccz_hosting_download_ccz', args=[
+                self.domain, self.file_hosting.id])
+        }
+
+    def store_file_in_blobdb(self, file_obj, name):
+        kw = {
+            "domain": self.file_hosting.domain,
+            "parent_id": 'CCZHosting',
```

Comment:
From a previous thread, which I can no longer reply to
> Reiterating on the why i cant use "CCZHosting-%s" % self.ccz_hosting.id" with the same blob id since that is not allowed due to unique constraint on blob id(key) for blob meta

I'm not following your reasoning for why `"CCZHosting-%s" % self.ccz_hosting.id"` will not work; there is no unique constraint on `parent_id`. Although I do see `self.file_hosting` is not always a `CCZHosting` object anymore.

The issue is that using a constant `parent_id` will cause metadata for all these blobs to be saved in the same database shard. That might not be a problem if there are only a few of them. But if it is expected that there will be many then it would be better to use a `parent_id` value that will cause the metadata to be distributed across shards. I do not know the best decision here since I don't know the expected count of these blobs.

Maybe `"%s-%s" % (type(self.file_hosting).__name__, self.file_hosting.id)` would work? `parent_id=self.blob_id` would certainly work, although that would make it harder to find the parent object given only the blob metadata.

Reply:
> there is no unique constraint on `parent_id`
@millerdev there is a unique constraint on `key` on BlobMeta and so i can't add blob meta again for a new ccz hosting in case the ccz for the same app + version + build profile is already present.

Reply:
> That might not be a problem if there are only a few of them

i am guessing if we consider only ICDS, and they launch in all 29 states with each link having the 3 applications or even if they follow some other convention i believe this would be < 100.

Reply:
> @millerdev there is a unique constraint on key on BlobMeta and so i can't add blob meta again for a new ccz hosting in case the ccz for the same app + version + build profile is already present.

There must be excactly one `BlobMeta` record per blob, no more and no less—the design of the blob db requires it. The issue you raise here should not be a concern since you are not reusing blob keys and it's not possible to have two blobs with the same key.

Reply:
> ... i believe this would be < 100

Yes, that's a very small number. Distribution among shards is not a concern if that is the max. We would need to revisit this decision if that ever changes.